### PR TITLE
Database: Use QueryScan rowFunc for SelectObjects

### PR DESCRIPTION
**Pull request by masnax**
This is a few additions to #10793.

* Updates `SelectObjects` to use the same callback function as `QueryScan`. I noticed a lot of anonymous structs floating around the schema updates. I converted these to type definitions to keep the consistency, but is there any reason we don't want to just call the database entities here? I can think maybe we don't want to import the `cluster` package yet, or perhaps a lot of these just predate those definitions? If it's the latter than I'll switch them where I can.

* Removes the `panic`s called on `Stmt` and `StmtString`. This means we had to handle the error that's returned, so the generator is updated to do that now, as well as other calls to `Stmt` in the code. 

* Moves `cluster_groups` to the `cluster` package and invokes `lxd-generate` for several of its functions. All of the ones that are now generated were using the exact same log as the generator already anyway. I had to make them play nice with `Stmt`'s new error, and so moving it over seemed like the better option.

### CODE REVIEW

**Comment by tomponline** [comment]
> * I noticed a lot of anonymous structs floating around the schema updates. I converted these to type definitions to keep the consistency, but is there any reason we don't want to just call the database entities here? I can think maybe we don't want to import the `cluster` package yet, or perhaps a lot of these just predate those definitions? If it's the latter than I'll switch them where I can.

If its just an internal detail and doesn't change the function's return type and there's an existing equivalent struct then I would say use it.

**Comment by tomponline** [comment]
> * exact same log

what does that mean?

File: lxd/db/storage_pools.go [hunk 1]
```diff
@@ -273,31 +277,27 @@ func (c *ClusterTx) CreatePendingStoragePool(node, name, driver string, conf map
...
 		return err
 	}
 
 	defer func() { _ = stmt.Close() }()
-	err = query.SelectObjects(stmt, dest, name)
+	count := 0
+	err = query.SelectObjects(stmt, func(scan func(dest ...any) error) error {
+		// Ensure that there is at most one pool with the given name.
+		if count != 0 {
+			return fmt.Errorf("more than one pool exists with the given name")
```
Comment (tomponline) [inline lxd/db/storage_pools.go:290]: Can we make sure that errors start with a capital letter please.

**Comment by masnax** [comment]
> > * exact same log
> 
> what does that mean?

sorry, autocorrect. It's logic*

File: lxd/db/networks.go [hunk 2]
```diff
@@ -299,32 +302,26 @@ func (c *ClusterTx) CreatePendingNetwork(node string, projectName string, name s
...
 	}
 
 	defer func() { _ = stmt.Close() }()
 
-	err = query.SelectObjects(stmt, dest, projectName, name)
+	count := 0
+	err = query.SelectObjects(stmt, func(scan func(dest ...any) error) error {
+		// Ensure that there is at most one network with the given name.
+		if count != 0 {
+			return fmt.Errorf("More than one network exists with the given name")
```
Comment (tomponline) [inline lxd/db/networks.go:316]: Can you use `return api.StatusErrorf(http.StatusConflict, "More than one network exists with the given name")` here?

Comment (tomponline) [reply lxd/db/networks.go:316]: also `count` is never incremented, so this can never be reached?

File: lxd/db/networks.go [hunk 3]
```diff
@@ -299,32 +302,26 @@ func (c *ClusterTx) CreatePendingNetwork(node string, projectName string, name s
...
 	stmt, err := c.tx.Prepare("SELECT id, state, type FROM networks WHERE project_id = (SELECT id FROM projects WHERE name = ?) AND name=?")
 	if err != nil {
 		return err
 	}
 
 	defer func() { _ = stmt.Close() }()
 
-	err = query.SelectObjects(stmt, dest, projectName, name)
+	count := 0
+	err = query.SelectObjects(stmt, func(scan func(dest ...any) error) error {
```
Comment (tomponline) [inline lxd/db/networks.go:313]: Why not use `query.QueryScan` here as the c.tx.Prepare call above is wasteful as its a one-time query.

File: lxd/db/storage_pools.go [hunk 4]
```diff
@@ -273,31 +277,27 @@ func (c *ClusterTx) CreatePendingStoragePool(node, name, driver string, conf map
...
-
 	stmt, err := c.tx.Prepare("SELECT id, driver, state FROM storage_pools WHERE name=?")
 	if err != nil {
 		return err
 	}
 
 	defer func() { _ = stmt.Close() }()
-	err = query.SelectObjects(stmt, dest, name)
+	count := 0
+	err = query.SelectObjects(stmt, func(scan func(dest ...any) error) error {
```
Comment (tomponline) [inline lxd/db/storage_pools.go:287]: Why don't we use query.QueryScan here instead of preparing a one-time query above?

**Comment by tomponline** [comment]
Please can you review the use of `SelectObjects` and if its being used with a one-time prepared statement, replace with `QueryScan` instead.

**Review by tomponline — CHANGES_REQUESTED** [review]
Looking good, a few logic bugs and some potential improvements to reduce the amount of one-time prepared statements.

**Comment by tomponline** [comment]
Tests failing.

**Comment by masnax** [comment]
The `clustering_groups` test was failing because the table `nodes_cluster_groups` wasn't being handled correctly with the update to `cluster_groups` to use lxd-generate. Now `nodes_cluster_groups` also uses lxd-generate, but to do so I had to add an `ID` column to that table, so there's a schema update in this PR now as well.

I also went ahead and cleared out a bunch of the `tx.Prepare`s floating around the codebase.

File: lxd/db/cluster/cluster_groups.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+//go:generate mapper method -i -e cluster_group Rename
+//go:generate mapper method -i -e cluster_group Create
+//go:generate mapper method -i -e cluster_group Update
+//go:generate mapper method -i -e cluster_group DeleteOne-by-Name
+
+// ClusterGroup is a value object holding db-related details about a cluster group.
+type ClusterGroup struct {
+	ID          int
+	Name        string
+	Description string   `db:"coalesce=''"`
```
Comment (tomponline) [inline lxd/db/cluster/cluster_groups.go:31]: Why `coalesce` here? This is a not null field and there are no joins I can see?

**Comment by tomponline** [comment]
jenkins: test this please

File: lxd/db/cluster/cluster_groups.go (hunk 5 above)
Comment (masnax) [reply lxd/db/cluster/cluster_groups.go:31]: The original query had it. Actually, it seems we do this for several `description` columns, though it doesn't look like any of those are ever null either. I'll make another PR where I'll clear them all out.

Comment (tomponline) [reply lxd/db/cluster/cluster_groups.go:31]: Thanks!

**Comment by tomponline** [comment]
I'm just re-running jenkins to check this builds after https://github.com/lxc/lxd/pull/10694 was merged.

**Review by tomponline — APPROVED** [review]
LGTM!

**Pull request merged**